### PR TITLE
Fixed #30578 - Made SelectDateWidget respect a custom date format when USE_L10N is disabled.

### PR DESCRIPTION
**Pull request by CruxBox**
Had reproduced and yielded invalid dates. Now fixed and updated tests.

### CODE REVIEW

**Comment by CruxBox** [comment]
I guess force merge will do @felixxm

File: django/forms/widgets.py [hunk 1]
```diff
@@ -1070,15 +1070,14 @@ def value_from_datadict(self, data, files, name):
...
-                except ValueError:
-                    pass
-                else:
-                    date_value = datetime_safe.new_date(date_value)
-                    return date_value.strftime(input_format)
+            input_format = get_format('DATE_INPUT_FORMATS')[0]
+            try:
+                date_value = datetime.date(int(y), int(m), int(d))
+            except ValueError:
+                pass
```
Comment (felixxm) [inline django/forms/widgets.py]: `return '%s-%s-%s' % (y or 0, m or 0, d or 0)` can be moved here.

File: django/forms/widgets.py [hunk 2]
```diff
@@ -1070,18 +1070,16 @@ def value_from_datadict(self, data, files, name):
...
+            input_format = get_format('DATE_INPUT_FORMATS')[0]
+            try:
+                date_value = datetime.date(int(y), int(m), int(d))
+            except ValueError:
+                # Return pseudo-ISO dates with zeros for any unselected values,
+                # e.g. '2017-0-23'.
+                return '%s-%s-%s' % (y or 0, m or 0, d or 0)
+            else:
+                date_value = datetime_safe.new_date(date_value)
+                return date_value.strftime(input_format)
```
Comment (charettes) [inline django/forms/widgets.py]: Minor but the `else` clause wrapping is unnecessary here given you return in the `except`.

Comment (CruxBox) [reply django/forms/widgets.py]: It is necessary. When try doesn't throw an exception, the else block is executed.

Comment (charettes) [reply django/forms/widgets.py]: @CruxBox my point is that you are already returning in the case of an exception *and* in the `else` clause so the code is equivalent to

```python
try:
    date_value = datetime.date(int(y), int(m), int(d))
except ValueError:
    return '%s-%s-%s' % (y or 0, m or 0, d or 0)
date_value = datetime_safe.new_date(date_value)
return date_value.strftime(input_format)
```

Again, this is minor but we usually avoid using `else` in the rest of the code base unless it's necessary.

Comment (CruxBox) [reply django/forms/widgets.py]: Oh right. Sorry I misinterpreted what you said earlier! Want me to remove the `else` then?

**Comment by felixxm** [comment]
@CruxBox Thanks :+1: I rebased from master and pushed minor edits.

**Comment by felixxm** [comment]
Merged in 26d16c07fdc4a297daca554afa6375c70d6d82a9.

**Pull request closed without merging**